blockchain/types/derivesha/mux.go

```diff
@@ -54,23 +59,32 @@ func init() {
 	}
 }
 
-func InitDeriveSha(chainConfig *params.ChainConfig) {
+func InitDeriveSha(chainConfig *params.ChainConfig, govEngine GovernanceEngine) {
 	config = chainConfig
+	gov = govEngine
 	types.DeriveSha = DeriveShaMux
 	types.EmptyRootHash = EmptyRootHashMux
 }
 
 func DeriveShaMux(list types.DerivableList, num *big.Int) common.Hash {
-	return instances[getType()].DeriveSha(list)
+	return instances[getType(num)].DeriveSha(list)
 }
 
 func EmptyRootHashMux(num *big.Int) common.Hash {
-	return emptyRoots[getType()]
+	return emptyRoots[getType(num)]
 }
 
-// TODO: Choose appropriate DeriveShaImpl from governance based on block number
-func getType() int {
+func getType(num *big.Int) int {
 	implType := config.DeriveShaImpl
+
+	if config.IsKoreForkEnabled(num) && gov != nil {
```

Comment:
I think `gov != nil` is enough. May I ask why you put the hardfork-enabled condition? Regardless of the hardfork activation, this new change seems necessary due to the fact that derive method can be dynamically changed. Plus, the absence of its condition works on a network that is not aware of Kore hardfork.

Reply:
Thank you for the input. I agree that `IsKore` condition is unnecessary.

<hr></hr>

My previous thought:

In a hypothetical setup with a v1.10 CN and a v1.9 EN,
- Kore Hardfork is scheduled at block 100. Istanbul.epoch is 30.
  - CN is upgraded to v1.10.
  - EN is not yet upgraded because hardfork has not arrived.
- CN, for some reason, votes `deriveshaimpl` at block 21.
- Block 30 contains governanceData with `deriveshaimpl`. The change will be applied from block 60.
- EN fails to ingest block 500 because of the difference between EN's changeset and block's governanceData. A bad block error occurs.

[Magma hardfork announcement](https://medium.com/klaytn/klaytn-v1-9-0-release-notes-medium-58e4644f7544) told node operators to upgrade their binary before the hardfork block. Similarly, I assume there will be v1.9 binaries operating before the Kore hardfork block.

The `IsKore` condition is unnecessary because
1. The condition in `getType()` does not protect bad block anyway. The bad block error is not from the discrepancy in DeriveShaImpl (from block 60), it's the discrepancy of governance data (from block 30).
2. It is unlikely that a v1.10 CN vote for `deriveshaimpl` before hardfork block.

Reply:
@hyunsooda Fixed in 6042a401. Thanks!

Reply:
Okay thanks checked.